Sources/AsyncKit/ConnectionPool/EventLoopGroupConnectionPool.swift

```diff
@@ -0,0 +1,188 @@
+        _ closure: @escaping (Source.Connection) -> EventLoopFuture<Result>
+    ) -> EventLoopFuture<Result> {
+        self.pool(for: eventLoop ?? self.eventLoopGroup.next())
+            .withConnection(logger: logger ?? self.logger, closure)
+    }
+    
+    /// Requests a pooled connection.
+    ///
+    /// The connection returned by this method should be released when you are finished using it.
+    ///
+    ///     let conn = try pool.requestConnection(...).wait()
+    ///     defer { pool.releaseConnection(conn) }
+    ///     // use the connection
+    ///
+    /// See `withConnection(_:)` for a callback-based method that automatically releases the connection.
+    ///
+    /// - parameters:
+    ///     - logger: For trace and debug logs.
+    ///     - eventLoop: Preferred event loop for the new connection.
+    /// - returns: A future containing the requested connection.
+    public func requestConnection(
+        logger: Logger? = nil,
+        on eventLoop: EventLoop? = nil
+    ) -> EventLoopFuture<Source.Connection> {
+        self.pool(for: eventLoop ?? self.eventLoopGroup.next())
+            .requestConnection(logger: logger ?? self.logger)
+    }
+    
+    /// Releases a connection back to the pool. Use with `requestConnection()`.
+    ///
+    ///     let conn = try pool.requestConnection(...).wait()
+    ///     defer { pool.releaseConnection(conn) }
+    ///     // use the connection
+    ///
+    /// - parameters:
+    ///     - connection: Connection to release back to the pool.
+    ///     - logger: For trace and debug logs.
+    public func releaseConnection(
+        _ connection: Source.Connection,
+        logger: Logger? = nil
+    ) {
+        self.pool(for: connection.eventLoop)
+            .releaseConnection(connection, logger: logger ?? self.logger)
+    }
+    
+    /// Returns the `EventLoopConnectionPool` for a specific event loop.
+    public func pool(for eventLoop: EventLoop) -> EventLoopConnectionPool<Source> {
+        self.storage[eventLoop.key]!
+    }
+    
+    /// Closes the connection pool.
+    ///
+    /// All available connections will be closed immediately.
+    /// Any connections currently in use will be closed when they are returned to the pool.
+    ///
+    /// Once closed, the connection pool cannot be used to create new connections.
+    ///
+    /// Connection pools must be closed before they deinitialize.
+    public func shutdown() {
+        // synchronize access to closing
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        // check to make sure we aren't double closing
+        guard !self.didShutdown else {
+            return
+        }
+        self.didShutdown = true
+        self.logger.debug("Connection pool shutting down, closing each event loop's storage")
+
+        // shutdown all pools
+        for pool in self.storage.values {
+            pool.shutdown()
+        }
+    }
+    
+    deinit {
+        if !self.didShutdown {
```

Comment:
Why not simply shut down automatically upon deinit? Also, you may want to synchronize access to `self.didShutdown` in this method with `lock`.

Reply:
This follows how NIO's types like `MultiThreadedEventLoopGroup` and `NIOThreadPool` work as well as SSWG's async-http-client's `HTTPClient`. Shutting down is synchronous (blocking), too so we need to be careful about where it happens.